src/VisualStudio/Core/Def/ErrorReporting/VisualStudioInfoBar.cs

```diff
@@ -4,119 +4,174 @@
-                            break;
-                        case InfoBarUI.UIKind.HyperLink:
-                            actionItems.Add(new InfoBarHyperlink(item.Title));
-                            break;
-                        case InfoBarUI.UIKind.Close:
-                            break;
-                        default:
-                            throw ExceptionUtilities.UnexpectedValue(item.Kind);
-                    }
-                }
-
-                var infoBarModel = new InfoBarModel(
-                    new[] { new InfoBarTextSpan(message) },
-                    actionItems,
-                    KnownMonikers.StatusInformation,
-                    isCloseButtonVisible: true);
-
-                var infoBarUI = factory.CreateInfoBar(infoBarModel);
-                if (infoBarUI == null)
-                    return;
-
-                uint? infoBarCookie = null;
-                var eventSink = new InfoBarEvents(items, onClose: () =>
-                {
-                    Contract.ThrowIfFalse(_threadingContext.JoinableTaskContext.IsOnMainThread);
-
-                    // Remove the message from the list that we're keeping track of.  Future identical
-                    // messages can now be shown.
-                    _currentlyShowingMessages.Remove(message);
-
-                    // Run given onClose action if there is one.
-                    items.FirstOrDefault(i => i.Kind == InfoBarUI.UIKind.Close).Action?.Invoke();
-
-                    if (infoBarCookie.HasValue)
-                    {
-                        infoBarUI.Unadvise(infoBarCookie.Value);
-                    }
-                });
-
-                infoBarUI.Advise(eventSink, out var cookie);
-                infoBarCookie = cookie;
-
-                infoBarHost.AddInfoBar(infoBarUI);
-
-                _currentlyShowingMessages.Add(message);
+                infoBarUI.Unadvise(infoBarCookie.Value);
             }
         });
+
+        if (ErrorHandler.Succeeded(infoBarUI.Advise(eventSink, out var cookie)))
+            infoBarCookie = cookie;
+
+        infoBarHost.AddInfoBar(infoBarUI);
+
+        _currentlyShowingMessages.Add(message);
+        return new InfoBarMessage(this, infoBarHost, message, imageMoniker, infoBarUI, infoBarCookie);
+    }
+
+    private object? GetInfoBarHostObject()
+    {
+        _threadingContext.ThrowIfNotOnUIThread();
+
+        if (_windowFrame != null)
+            return ErrorHandler.Succeeded(_windowFrame.GetProperty((int)__VSFPROPID7.VSFPROPID_InfoBarHost, out var windowInfoBarHostObject)) ? windowInfoBarHostObject : null;
+
+        return _serviceProvider.GetService(typeof(SVsShell)) is IVsShell shell && ErrorHandler.Succeeded(shell.GetProperty((int)__VSSPROPID7.VSSPROPID_MainWindowInfoBarHost, out var globalInfoBarHostObject))
+            ? globalInfoBarHostObject
+            : null;
+    }
+
+    public sealed class InfoBarMessage(
```

Comment:
rich object returned back to represent the message being shown to the user.  Used by the sg system as it wants to programatically remove the message before updating it to new info.